text/2021-04-23-builtin-gc-worker.md

```diff
@@ -0,0 +1,169 @@
+# Builtin GC worker
+
+## Summary
+
+1. Move TiKV MVCC GC worker from TiDB into PD and make it general for use without TiDB.
+
+2. Support different safepoint for various ranges.
+
+## Motivation
+
+1. GC worker is an important component for TiKV that deletes outdated MVCC data so as to not explode the storage. But currently, the GC worker is implemented in TiDB, which makes TiKV not usable without TiDB.
+
+2. Flashback is a feature under design for TiDB which allows querying SQL using a specified timestamp. By specifying GC deadline for different tables we can enlarge the use case that allowing arbitrary early timestamp (currently 10 mins) in flashback as long as a proper GC deadline is set for the table.
+
+## Background
+
+According to the [documentation](https://docs.pingcap.com/tidb/stable/garbage-collection-overview) for the **current** GC worker, the GC worker processes the following steps in a GC cycle:
+
+1. Calculate the minimal timestamp among all living SQL sessions, earlier than which are safe to GC, called `Session Safepoint`. Then calculate the `Service Safepoint` for id `gc_worker` by `min(Session Safepoint, now - gc_life_time)`.
+2. [Push](https://github.com/pingcap/kvproto/blob/8ecb5e46d7f5f7952a1a8d262b54f61dc8de1ef3/proto/pdpb.proto#L73) the `gc_worker Service Safepoint` to PD. At the meanwhile, other tools (e.g. CDC, BR) has also pushed their `Service Safepoint` for minimal snapshot timestamp requirement.
+3. Get the minimal `Service Safepoint` among all services from the response of step 2, which is `GC Safepoint`.
+4. Resolve all locks whose timestamp is earlier than the `GC Safepoint`.
+5. [Push](https://github.com/pingcap/kvproto/blob/8ecb5e46d7f5f7952a1a8d262b54f61dc8de1ef3/proto/pdpb.proto#L71) the `GC Safepoint` to PD.
+
+After the `GCSafepoint` being uploaded, TiKV will pull it from PD by interval, then TiKV will automatically delete all records earlier than `GCSafepoint`.
+
+Yet another tricky part is that TiDB has a shortcut to GC a range of keys after `DROP TABLE` or `DROP INDEX`, etc: parallel with step 3 and step 4, TiDB will directly invoke `UnsafeDeleteRange` to TiKV which will delete the range of key-values regardless of locks or timestamp. To make it safe, TiDB will guarantee that the deleted range will never be used again.
+
+## Detailed design
+
+Four new concepts will be introduced by this proposal:
+
+1. `Safeline`
```

Comment:
Personally I prefer still naming it *safepoint* since it's still a point in the timestamp axis, and it's a term that has been used for years by us, the uses and DBAs.